### ISSUE TITLE
Auto-detect Scala file changes

### ISSUE THREAD

**Issue by gvolpe** [VirtusLab/scala-cli#1962]
**Version(s)**
Tested with v0.2.1. 

**Describe the bug**
I had a chat with @tgodzik last Friday at the Scalar conference, and I showed him an [example](https://github.com/gvolpe/scalar-feda) that failed to integrate with Metals, and he was kind enough to help me figure out a workaround. Not sure whether this belongs here or in the Metals repo, happy to move it :)

**To Reproduce**
The issue begins when one creates a simple `main.scala` file and runs `scala-cli run main.scala`. The `scala-cli` program would generate a `.bsp/scala-cli.json` configuration file on the first run (I think). This works well and integrates just fine with Metals.

Now let's say we split up that initial `main.scala` into multiple Scala files (that's the process I went through in the example linked above). Running `scala-cli run .` will continue to work, but now the Metals integration is broken.
 
To fix the issue, we need to manually run `scala-cli setup-ide .`, so that the `.bsp/scala-cli.json` is generated once more with the correct configuration. After restarting Metals, it should resume working.
 
**Expected behaviour**
Tomasz mentioned that this UX could potentially be improved by auto-detecting file changes (perhaps, every time `scala-cli` runs?), so that the user doesn't have to figure out internal BSP configuration details, resulting in a better user experience.

Happy to provide more details, in case I missed something in this report. Thanks again for making `scala-cli`, it rocks! 🤘🏽

**Comment by Gedochao** [comment]
The current intended behaviour is that the IDE import data (the `.bsp/scala-cli.json` BSP connection file) is generated the first time `run`, `compile` or `test` sub-commands are called, or whenever `setup-ide` is called.
The default sub-command with passed inputs counts as `run` for this purpose.

What this means that subsequent calls of `run`/`compile`/`test` sub-commands in the same directory will not change the existing `.bsp/scala-cli.json` and it's the intended behaviour right now. 
`setup-ide` should be used to update the project structure.
```bash
scala-cli run main.scala # this is the first call of run, so the .bsp/scala-cli.json file is created
scala-cli run . # .bsp/scala-cli.json already exists, so calling run won't change it
scala-cli setup-ide . # this will override the currently existing .bsp/scala-cli.json
```

I don't think we should change this on Scala CLI's side, since the user might want to `run`/`compile`/`test` different sets of files, scopes and/or singular scripts, while not changing the project structure he has currently open in his IDE.
i.e. we don't want the following to change the project structure:
```bash
scala-cli run scripts/some-script.sc
```

Creating the BSP connection on the first run is merely for convenience, so that the user can open the IDE in his working directory seamlessly after running a command.
However, to control the project structure later, it's safer to rely on the `setup-ide` sub-command, where the intention of the user is clear.

The downside is indeed the scenario described in this issue, where the user first runs some single file, and then wants to add files to it without prior knowledge of this mechanism or the `setup-ide` command. 
Long story short, we trust the user knows what he's doing, and he might not, in fact, know what he's doing.

However, in this scenario maybe it'd be possible to interactively ask the user if he/she wants to include more files to the project scope on the IDE's side, if they are present in the project root directory... (?)
cc @tgodzik

**Comment by ckipp01** [comment]
Just to tie these together. I actually created an issue for this on the Metals side quite some time ago. https://github.com/scalameta/metals/issues/4741. I sort of agree here that `setup-ide` shouldn't need to be called again. If another file is added to a build target, `scala-cli` should know that. If this happens, could something like a build target changed notification be used to let the client know that a new file has been added to the target?

**Comment by gvolpe** [comment]
@Gedochao thanks a lot for your comment! 

I think it makes a lot of sense to leave it as is 👍🏽 It's even [documented](https://scala-cli.virtuslab.org/docs/guides/ide#directories-vs-single-files-when-working-with-an-ide), so can't expect more 😄

**Comment by gvolpe** [comment]
>  If this happens, could something like a build target changed notification be used to let the client know that a new file has been added to the target?

Thanks Chris! That could certainly improve the UX 👍🏽

**Comment by Gedochao** [comment]
> I sort of agree here that setup-ide shouldn't need to be called again.

I mean, it's not strictly called **again**... 
Technically, creating the BSP connection json (IDE import data) upon the first `run`/`compile`/`test` call is a side effect, most users wouldn't even be aware this happens implicitly.
It's `run`/`compile`/`test`/`setup-ide` for first setup of a project to be imported to an IDE, and then `setup-ide` for further control of the project scope.

> If this happens, could something like a build target changed notification be used to let the client know that a new file has been added to the target?

Except we don't know if something has been added to the build target, for all we know it's a new build target. 
Subsequent `run` calls might be totally unrelated, like running a script, another main, another scope, etc. 
Passing `.` to a command is the default, but people may choose to pass anything as an input, really (multiple inputs, even).

```bash
scala-cli main.scala 
scala-cli main2.scala # is main2.scala added to the build target, do we have a new build target? who knows
scala-cli . # ah, so now we're running the whole directory, are we?
scala-cli test testSources 
scala-cli some-script.sc
```

We kinda guess that the first time `run`/`compile`/`test` is called in a directory, it's fair to assume the user might want to import it to his IDE.
However, with subsequent calls it's hard to keep guessing the intentions of the user, so we just don't.
Explicitly calling `setup-ide` makes that intention clear.

Ultimately, Scala CLI has a command-line first approach. 
However, I do think it could be done on the IDE's side instead.
I don't think the IDE should include new stuff in the build scope arbitrarily, but an interactive pop-up could help, maybe...
Something like:
```
Hey, it seems the project root directory contains files which aren't included in your Scala CLI build target. Would you like to add them?
Yes/No
```
In case of `yes`, `scala-cli setup-ide .` could even be called by the IDE, theoretically. 
Any options passed to `setup-ide` are saved in `.scala-build/ide-options-v2.json`

**Comment by ckipp01** [comment]
Ah so I think I misunderstood since I just tested this and I was under the assumption that if you did:

```
scala-cli setup-ide .
```
and then created a new file, it wouldn't be recognized, but it is actually. So I guess in theory that does make sense, but I can also see how that may totally bite someone if they do `scala-cli setup-ide main.scala` and then add a `Foo.scala` later and wonder why it doesn't work.

**Comment by gvolpe** [comment]
Thanks everyone for sharing your inputs, closing this 👍🏽
